libraries/postgres/src/Database.js

```diff
@@ -721,6 +744,49 @@ class Database {
 
     return Buffer.concat([b1, b2]);
   }
+
+  /**
+   * Given a row and an array of serialization functions, generate a signature
+   * container using the last serialization in the array.
+   */
+  sign({row, signingSerializations}) {
+    const ser = signingSerializations.length - 1;
+    const v = 0;
+    const {id: kid} = this.signingKeyring.currentKey('hmac-sha512');
+    const sig = this.calculateSignature({row, kid, v, ser, signingSerializations});
+    return {v, ser, kid, sig};
+  }
+
+  /**
+   * Given a row, signature, and serializations, verify the signature.  This
+   * performs a fixed-time comparison to avoid timing attacks on the signature.
+   * Returns true if the signature is valid.
+   */
+  verifySignature({row, signature, signingSerializations}) {
+    const sig = this.calculateSignature({
+      row,
+      kid: signature.kid,
+      v: signature.v,
+      ser: signature.ser,
+      signingSerializations,
+    });
+    return fixedTimeComparison(Buffer.from(signature.sig, 'base64'), Buffer.from(sig, 'base64'));
```

Comment:
I'm not sure I understand everything with fixed time comparison. Do you have a link to suggest that I can use to understand this better? Thanks!

Reply:
https://severalnines.com/database-blog/utilizing-encryption-strengthen-postgresql-database-security is a good overview of postgres's extensive security features, and https://research.kudelskisecurity.com/2013/12/13/timing-attacks-part-1/ is a specific article about timing attacks.

Reply:
To the first link, note that we are specifically *not* doing any crypto on the database itself.  All encryption and signing takes place in Node, with keys and plaintext never visible on the DB server.

In this context I doubt a timing attack would be practical, since the timing of the comparison would be miniscule copmared to the DB query, network traffic, multiple layers of HTTP proxying, calls out to auth.authenticateHawk, and k8s and virutalization overhead.  But it's a best practice and it's easy, so we do it.

Reply:
Unfortunately that is (understandably) a common misconception. To quote the conclusion of the above cited study:

<img width="407" alt="Screenshot 2020-07-03 at 17 35 46" src="https://user-images.githubusercontent.com/190790/86483170-c0096d80-bd53-11ea-89a9-33eba2b20bec.png">

So you are right that a lot of effort would be needed to launch such an attack successfully. But the key result of this paper is that it isn't infeasible, and such attacks have been successful against other targets in the past, since even a considerable amount of external noise can be filtered out if enough data is collected.

Reply:
Yep, timing attacks are the basis of meltdown and spectre, too.